### DISCUSSION TITLE
Explanation on decimals in solidity

### DISCUSSION THREAD

**Discussion by EddiePumpin** [Cyfrin/foundry-full-course-cu#2873] [Q&A] [answered]
```  ///////////////////
    // State Variables
    ///////////////////
    uint256 private constant ADDITIONAL_FEED_PRECISION = 1e10;
    uint256 private constant PRECISION = 1e18;```

```function getUsdValue(address token, uint256 amount) public view returns (uint256) {
            AggregatorV3Interface priceFeed = AggregatorV3Interface(s_priceFeeds[token]); // See it as more or less like creating a new instance
            (, int256 price,,,) = priceFeed.latestRoundData(); // The price is the 1 of the token in USD 
            return ((uint256(price) * ADDITIONAL_FEED_PRECISION) * amount)/PRECISION; // Price is multiplied with 10e8 because the decimal of ETH/USD is 8
        }```
        
        Please, I need a detailed explanation on this decimal of a thing. I have tried my best to understand but it is not working.

**Comment by LordAdhaar** [comment]
What is the error you are facing? Or what is that you are not able to understand? If you give more context I will be able to help you out.

**Reply by EddiePumpin** [reply]
Yeah, I didn't understand the conversion.

**Comment by EngrPips** [comment]
Hello @EddiePumpin, Please provide more details as @LordAdhaar has stated.

**Comment by satyaorz** [comment, marked as answer]
The thing is,

 when you do
 
``AggregatorV3Interface priceFeed = AggregatorV3Interface(s_priceFeeds[token]);
     (, int256 price,,,) = priceFeed.latestRoundData();``
     
the ``price`` that ``priceFeed.latestRoundData()`` returns is the USD price of 1 ETH scaled by 1e8 to handle the faction part of the actual price of ETH in USD

imagine 1 ETH price in USD is 2456.567
then ``int256 price`` will be equal to 2456``56700000``, if it was not scaled by ``1e8``, it would be a problem, as solidity does not have ``float`` variable type

for example,
imagine 1 ETH price in USD is 2456$ so ``int256 price`` will be equal to ``245600000000``, and you want to check what USD worth of Ether the sender is sending, but the sender is not sending ETH, they are sending wei instead;

so if you wanted to check the value of 1 ETH in terms of wei

as 1 ETH = 1000000000000000000 or 1e18 wei

you have to scale the ```price``` by ``1e10`` more, as the price of ETH is already scaled by ``1e8``


when you are doing ``(uint256(price) * ADDITIONAL_FEED_PRECISION)``
in our example
it does ``245600000000 * 1e10`` which is equals  = 2456000000000000000000

and if the sender is sending lets say ``100000000000000000 or 1e17`` wei which is 0.1 ETH
it will  be multiplied to ``2456000000000000000000`` 

and it becomes ``245600000000000000000000000000000000000`` or ``2456000000000000000000 * 1e17`` 
and now if you devide it by PRECISION which is 1e18

you get 245600000000000000000 which is the price of 0.1 ETH or 1e17 wei
which is 245.6 * 1e8 * 1e10

which equal to 245.6 USD in terms of wei

**Reply by EngrPips** [reply]
Great explanation here @EddiePumpin

**Reply by LordAdhaar** [reply]
Good name, I am hungry now

**Reply by EddiePumpin** [reply]
So, by default the returned price is scaled by 1e8?

**Reply by satyaorz** [reply]
Yes, so that it can handle fraction part of the price. 
Like I said, if the price was 2456.567 you cannot represent it in solidity so it's scaled by 1e8 to give you 245656700000 , then further as the  sender is sending funds in Wei, the price is scaled by 1e10 more, total scaled 1e18.

**Reply by EngrPips** [reply]
Interesting

**Reply by EddiePumpin** [reply]
Thanks man. I understand now but there is a bug with that code.

![Screenshot (38)](https://github.com/user-attachments/assets/bdc05e4a-6a47-408d-a6ba-c8f0d89ea0f5)

**Reply by EngrPips** [reply]
Awesome.
